### PR TITLE
Task/07 api

### PR DESCRIPTION
**Pull request by greyzerua**
**Acceptance Criteria for 1**

- [x]  Mentor is added to PR
- [x]  AC copied to PR description
- [x]  Link to deployed version is added to PR https://greyzerua.github.io/fe-react-2024/
- [x]  ProductPage fetches data from the API using either fetch or axios
- [x]  Pagination is implemented using query parameters
- [x]  Search functionality is implemented using query parameters, with the search triggered by a search button
- [x]  Category filtering is implemented using query parameters, resetting the page to the first page on change
- [x]  A loading spinner or message is displayed while awaiting API requests
- [x]  If the API response is an empty array or an error occurs, a "No products" message is displayed instead of a blank screen

**Acceptance Criteria for 1***

- [x]  Pagination is hidden on mobile view
- [x]  Infinite scroll is implemented for mobile view, either manually or using a library
- [x]  Error notifications are implemented using axios interceptors, with component libraries for toastr notifications

photo_2024-06-15_13-11-31

### CODE REVIEW

File: src/pages/product-page/hooks/useProduct.ts [hunk 1]
```diff
@@ -1,24 +1,26 @@
...
-                .then((productResponse) => {
-                    setProduct(productResponse);
-                });
+        async (productId: number) => {
+            try {
+                const result = await ApiService.GetInstance().get<Product>(`products/${productId}`);
+                setProduct(result);
+            } catch (error: unknown) {
+                const axiosError = ApiService.IsAxiosError(error);
+                if (axiosError?.response?.status === 400) {
```
Comment (vitalii-mel-work) [inline src/pages/product-page/hooks/useProduct.ts]: status code to enum

File: src/pages/products-page/hooks/useGetProducts.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    const checkIsInfinite = () => {
+        setIsInfinite(window.innerWidth <= INFINITE_SCROLL_WINDOW_WIDTH);
+    };
+
+    useEffect(() => {
+        checkIsInfinite();
+        window.addEventListener('resize', checkIsInfinite);
+        return () => {
+            window.removeEventListener('resize', checkIsInfinite);
```
Comment (vitalii-mel-work) [inline src/pages/products-page/hooks/useGetProducts.ts]: create custom hook with params callback

File: src/pages/products-page/hooks/useGetProducts.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        };
+    }, []);
+
+    const queryParameters = useMemo(() => {
+        const trimmedTitle = searchValue.trim();
+        const parameters = {
+            ...(categories === null ? {} : { categoryId: String(categories) }),
+            ...(trimmedTitle.length > 0 ? { title: trimmedTitle } : {}),
```
Comment (vitalii-mel-work) [inline src/pages/products-page/hooks/useGetProducts.ts]: !trimmedTitle.length

File: src/pages/products-page/products-page.tsx [hunk 4]
```diff
@@ -1,68 +1,72 @@
...
+                    loader={<Icon iconType={EIconType.LOADING} width={70} height={70} className={styles['products__loading']} />}
+                >
+                    <ProductsList products={data} />
+                </InfiniteScroll>
+            );
+        }
+        if (isLoading) {
+            return <Icon iconType={EIconType.LOADING} className={styles['products__loading']} />;
+        }
+        if (data.length === 0) {
```
Comment (vitalii-mel-work) [inline src/pages/products-page/products-page.tsx:38]: !!data.length

**Review by vitalii-mel-work — APPROVED** [review]

File: src/pages/product-page/hooks/useProduct.ts (hunk 1 above)
Comment (greyzerua) [reply src/pages/product-page/hooks/useProduct.ts]: fixed

File: src/pages/products-page/hooks/useGetProducts.ts (hunk 2 above)
Comment (greyzerua) [reply src/pages/products-page/hooks/useGetProducts.ts]: fixed

File: src/pages/products-page/hooks/useGetProducts.ts (hunk 3 above)
Comment (greyzerua) [reply src/pages/products-page/hooks/useGetProducts.ts]: fixed

File: src/pages/products-page/products-page.tsx (hunk 4 above)
Comment (greyzerua) [reply src/pages/products-page/products-page.tsx:38]: fixed

**Review by vitalii-mel-work — APPROVED** [review]

**Pull request merged**